File: generators/cs/CRC16.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace HashFunctions
{
    public class CRC16 : IHashFunction
    {
        public override string Digest(byte[] input)
        {
            List<UInt16> message = new List<UInt16>();

            for (int i = 0; i < input.Length; i++)
            {
                message.Add(Convert.ToUInt16(input[i]));
            }

            UInt16 crc = 0xFFFF;

            for (int i = 0; i < message.Count; i++)
            {
                UInt16 ch = message[i];

                for (int j = 0; j < 8; j++) {
                    UInt16 b = (UInt16)((ch ^ crc) & 1);

                    crc >>= 1;

                    if (b == 1) {
                        crc = (UInt16)(crc ^ 0x8408);
                    }

                    ch >>= 1;
                }
            }

            string res = (~crc).ToString("x4");

            return res.Substring(res.Length - 4);
        }
    }
}

File: generators/cpp/crc16.cpp
#include "crc16.h"

using namespace std;

string CRC16::getHex(unsigned long in) {
    bitset<16> bs(in);
    unsigned n = bs.to_ulong();

    stringstream sstream;
    sstream << std::hex << std::setw(4) << std::setfill('0') << n;
    string temp;
    sstream >> temp;

    return temp;
}

string CRC16::digest(string input)
{
    unsigned short crc = 0xFFFF;
    unsigned char ch, b;
    int s = input.size();

    for (int i = 0; i < s; i++) {
        ch = input[i];

        for (int j = 0; j < 8; j++) {
            b = (ch ^ crc) & 1;
            crc >>= 1;
            if (b) {
                crc = crc ^ 0x8408;
            }
            ch >>= 1;
        }
    }

    return getHex(~crc);
}